hedera-node/src/main/java/com/hedera/services/records/ConsensusTimeTracker.java

```diff
@@ -0,0 +1,312 @@
+ */
+@Singleton
+public class ConsensusTimeTracker {
+	private static final Logger log = LogManager.getLogger(ConsensusTimeTracker.class);
+
+	@VisibleForTesting
+	static final long DEFAULT_NANOS_PER_ROUND = TransactionHandler.MIN_TRANS_TIMESTAMP_INCR_NANOS;
+
+	/**
+	 * The max number of records that can be placed before the first transaction in a round. We need this because
+	 * the first transaction might have more than a normal number of proceeding records.
+	 * For example {@link com.hedera.services.state.migration.MigrationRecordsManager}
+	 */
+	public static final long MAX_PRECEDING_RECORDS_FIRST_TXN_IN_ROUND = 10;
+
+	/**
+	 * The max number of records that can be placed before subsequent transactions in a round.
+	 */
+	public static final long MAX_PRECEDING_RECORDS_REMAINING_TXN = 3;
+	/**
+	 * The max number of records that can be placed after a transaction.
+	 */
+	public static final long MAX_FOLLOWING_RECORDS = 10;
+
+	private final long nanosPerRound;
+
+	private Instant roundStartConsensusTime;
+	private Instant currentTxnMinTime;
+	private Instant currentTxnTime;
+	private Instant currentTxnMaxTime;
+	private Instant maxConsensusTime;
+
+	private long followingRecordsCount;
+
+	private boolean firstUsed;
+
+	@Inject
+	public ConsensusTimeTracker() {
+		this(DEFAULT_NANOS_PER_ROUND);
+	}
+
+	@VisibleForTesting
+	ConsensusTimeTracker(final long nanosPerRound) {
+		this.nanosPerRound = nanosPerRound;
+		if (nanosPerRound <
+				(MAX_PRECEDING_RECORDS_FIRST_TXN_IN_ROUND + MAX_FOLLOWING_RECORDS
+						+ MAX_PRECEDING_RECORDS_REMAINING_TXN + 10)) {
+			throw new IllegalArgumentException("TransactionHandler.MIN_TRANS_TIMESTAMP_INCR_NANOS is too small!");
+		}
+	}
+
+	@VisibleForTesting
+	ConsensusTimeTracker(ConsensusTimeTracker toCopy) {
+		this();
+		roundStartConsensusTime = toCopy.roundStartConsensusTime;
+		currentTxnMinTime = toCopy.currentTxnMinTime;
+		currentTxnTime = toCopy.currentTxnTime;
+		currentTxnMaxTime = toCopy.currentTxnMaxTime;
+		maxConsensusTime = toCopy.maxConsensusTime;
+		followingRecordsCount = toCopy.followingRecordsCount;
+		firstUsed = toCopy.firstUsed;
+	}
+
+	/**
+	 * @param count the actual number of records that were placed following the consensusTime of the current transaction.
+	 */
+	public void setActualFollowingRecordsCount(final long count) {
+		if (count < 0) {
+			throw new IllegalArgumentException("count must be >= 0, it was " + count);
+		}
+		followingRecordsCount = count;
+	}
+
+	/**
+	 * @return true if firstTransactionTime() cannot be used again in this round
+	 */
+	public boolean isFirstUsed() {
+		return firstUsed;
+	}
+
+	/**
+	 * Resets this tracker, should be called at the beginning of
+	 * {@link com.hedera.services.state.logic.StandardProcessLogic#incorporateConsensusTxn}
+	 * @param consensusTime the consensusTime passed to incorporateConsensusTxn
+	 */
+	public void reset(final Instant consensusTime) {
+		roundStartConsensusTime = consensusTime;
```

Comment:
In general, the consensus time from `incorporateConsensusTxn` will _not_ be the first in the round. (Rounds/sec is ~1/sec given typical network conditions, I think.)

Reply:
renaming field